### PR TITLE
ENG-820 provision users without a password

### PR DESCRIPTION
**Pull request by tonisefton**
This change is [Reviewable](https://reviewable.io/reviews/work180/auth0-management/3)

### CODE REVIEW

File: src/client.rs [hunk 1]
```diff
@@ -39,12 +45,18 @@ impl Auth0Client {
...
       let body = std::str::from_utf8(&body).unwrap();
-      let body = if body.is_empty() { "null" } else { body };
 
-      Ok(serde_json::from_str::<R>(body)?)
+      let body: &str = if body.is_empty() { "null" } else { body };
+
+      Ok( if res_is_json { serde_json::from_str::<R>(body)? } else {
+        let message = Auth0MessageResponse {message: body.to_string()};
+        serde_json::from_str::<R>(&*serde_json::to_string(&message)?)?
+      } )
```
Comment (tonisefton) [inline src/client.rs]: Bit hacky, open to options

**Review by mterring — CHANGES_REQUESTED** [review]
> Reviewed 1 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/work180/auth0-management/3)* status: 1 of 4 files reviewed, 3 unresolved discussions (waiting on @i-mann and @tonisefton)

___
*[`src/client.rs` line 3 at r1](https://reviewable.io/reviews/work180/auth0-management/3#-NFMm8r04IVXItfle-x_:-NFMm8r04IVXItfle-xa:bm6db3l) ([raw file](https://github.com/work180/auth0-management/blob/7dde14d18573af430b96192339be2c220f33e271/src/client.rs#L3)):*
> ```XML
> //! Auth0 request client.
> use reqwest::{Client, Method, RequestBuilder};
> use reqwest::header::CONTENT_TYPE;
> ```

Nit: merge with existing use statement (same below)
___
*[`src/client.rs` line 49 at r1](https://reviewable.io/reviews/work180/auth0-management/3#-NFMmsBs7YjN3t_3QRqD:-NFMmsBs7YjN3t_3QRqE:b-gpa1en) ([raw file](https://github.com/work180/auth0-management/blob/7dde14d18573af430b96192339be2c220f33e271/src/client.rs#L49)):*
> ```XML
>     if res.status().is_success() {
>       let headers = res.headers().to_owned();
>       let res_is_json = headers.contains_key(CONTENT_TYPE) && headers[CONTENT_TYPE] == "application/json; charset=utf-8";
> ```

Type safe way: https://docs.rs/reqwest/0.6.0/reqwest/mime/struct.Mime.html
___
*[`src/client.rs` line 59 at r1](https://reviewable.io/reviews/work180/auth0-management/3#-NFM4lHSrGI-0k0ShCr7-r1-59:-NFN4mTl4CnWL8ABpB47:b4dxt1s) ([raw file](https://github.com/work180/auth0-management/blob/7dde14d18573af430b96192339be2c220f33e271/src/client.rs#L59)):*
<details><summary><i>Previously, tonisefton (Toni Sefton) wrote…</i></summary><blockquote>

Bit hacky, open to options
</blockquote></details>

How about this - which could be called with `R` of `&str` (I think - `String` should def work) and removes the need for a dummy response type



_Suggestion:_
```XML
      let body: &str = if body.is_empty() { "null" } else if res_is_json { body } else { serde_json::to_string(body) };

      Ok(serde_json::from_str::<R>(body)?)
```

**Review by tonisefton — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/work180/auth0-management/3)* status: 0 of 4 files reviewed, 1 unresolved discussion (waiting on @i-mann, @mterring, and @tonisefton)

___
*[`src/client.rs` line 3 at r1](https://reviewable.io/reviews/work180/auth0-management/3#-NFMm8r04IVXItfle-x_:-NFOK3i66RBLlBkJ0PxA:b8swspm) ([raw file](https://github.com/work180/auth0-management/blob/7dde14d18573af430b96192339be2c220f33e271/src/client.rs#L3)):*
<details><summary><i>Previously, mterring (Michael Terrington) wrote…</i></summary><blockquote>

Nit: merge with existing use statement (same below)
</blockquote></details>

actioned
___
*[`src/client.rs` line 49 at r1](https://reviewable.io/reviews/work180/auth0-management/3#-NFMmsBs7YjN3t_3QRqD:-NFOK6AMCMUYI2JmsT9d:b-als9nx) ([raw file](https://github.com/work180/auth0-management/blob/7dde14d18573af430b96192339be2c220f33e271/src/client.rs#L49)):*
<details><summary><i>Previously, mterring (Michael Terrington) wrote…</i></summary><blockquote>

Type safe way: https://docs.rs/reqwest/0.6.0/reqwest/mime/struct.Mime.html
</blockquote></details>

I tried this but couldn't work out the imports it looks like Mime is from an older version of reqwest and I couldn't find anything in the latest documentation on this.
___
*[`src/client.rs` line 59 at r1](https://reviewable.io/reviews/work180/auth0-management/3#-NFM4lHSrGI-0k0ShCr7-r1-59:-NFOKasNDnJzVvpKeJaj:bzb9ssl) ([raw file](https://github.com/work180/auth0-management/blob/7dde14d18573af430b96192339be2c220f33e271/src/client.rs#L59)):*
<details><summary><i>Previously, mterring (Michael Terrington) wrote…</i></summary><blockquote>

How about this - which could be called with `R` of `&str` (I think - `String` should def work) and removes the need for a dummy response type
</blockquote></details>

This works! Thank you

**Review by mterring — COMMENTED** [review]
> Reviewed 3 of 4 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/work180/auth0-management/3)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @i-mann)

___
*[`src/client.rs` line 49 at r1](https://reviewable.io/reviews/work180/auth0-management/3#-NFMmsBs7YjN3t_3QRqD:-NFfToYX7bBAdLK7T0Ge:b4vqq96) ([raw file](https://github.com/work180/auth0-management/blob/7dde14d18573af430b96192339be2c220f33e271/src/client.rs#L49)):*
<details><summary><i>Previously, tonisefton (Toni Sefton) wrote…</i></summary><blockquote>

I tried this but couldn't work out the imports it looks like Mime is from an older version of reqwest and I couldn't find anything in the latest documentation on this.
</blockquote></details>

:man_facepalming: Gets me everytime...
___
*[`src/client.rs` line 47 at r2](https://reviewable.io/reviews/work180/auth0-management/3#-NFfVjJP8DtwKcKTMr0E:-NFfVjJP8DtwKcKTMr0F:bwiilfz) ([raw file](https://github.com/work180/auth0-management/blob/c4a31b0718dfa93e741fddffd49694310c3e1b73/src/client.rs#L47)):*
> ```XML
>     if res.status().is_success() {
>       let headers = res.headers().to_owned();
>       let res_is_json = headers.contains_key(CONTENT_TYPE) && headers[CONTENT_TYPE] == "application/json; charset=utf-8";
> ```

I believe this can be simplified:



_Suggestion:_
```XML
      let res_is_json = res.headers().get(CONTENT_TYPE).unwrap_or_default() == "application/json; charset=utf-8";
```

**Review by i-mann — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/work180/auth0-management/3)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @tonisefton)

___
*[`src/client.rs` line 47 at r2](https://reviewable.io/reviews/work180/auth0-management/3#-NFfVjJP8DtwKcKTMr0E:-NFlifGoAvrG81QxxJtR:b-binn24) ([raw file](https://github.com/work180/auth0-management/blob/c4a31b0718dfa93e741fddffd49694310c3e1b73/src/client.rs#L47)):*
<details><summary><i>Previously, mterring (Michael Terrington) wrote…</i></summary><blockquote>

I believe this can be simplified:
</blockquote></details>

Nope - trait Default is not implemented for &Header

**Review by mterring — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/work180/auth0-management/3)* status: all files reviewed, 1 unresolved discussion (waiting on @tonisefton)

___
*[`src/client.rs` line 47 at r2](https://reviewable.io/reviews/work180/auth0-management/3#-NFfVjJP8DtwKcKTMr0E:-NFmTHY9--PW73K2-bC9:bq3m3zz) ([raw file](https://github.com/work180/auth0-management/blob/c4a31b0718dfa93e741fddffd49694310c3e1b73/src/client.rs#L47)):*
<details><summary><i>Previously, i-mann (Isaac Mann) wrote…</i></summary><blockquote>

Nope - trait Default is not implemented for &Header
</blockquote></details>

Sigh... I thought headers returned a string. At least I think remove the `to_owned()` (afiak it's not required) or try this:



_Code snippet:_
```XML
let res_is_json = res.headers().get(CONTENT_TYPE).filter(|h| h == "application/json; charset=utf-8").is_some();
```
___
*[`src/client.rs` line 56 at r3](https://reviewable.io/reviews/work180/auth0-management/3#-NFmWr69-WvXd525EcXJ:-NFmWr69-WvXd525EcXK:b-16eb54) ([raw file](https://github.com/work180/auth0-management/blob/1399cd767c360c0f58e9e0ed0437e3ee0d2e88a4/src/client.rs#L56)):*
> ```XML
>       let body = std::str::from_utf8(&body).unwrap();
> 
>       let json_body = serde_json::to_string(body)?;
> ```

I don't think you want the cost of json encoding the body unless you really need it (which is the not-normal path for this code)

**Review by i-mann — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/work180/auth0-management/3)* status: all files reviewed, 1 unresolved discussion (waiting on @mterring and @robmarto)

___
*[`src/client.rs` line 47 at r2](https://reviewable.io/reviews/work180/auth0-management/3#-NFfVjJP8DtwKcKTMr0E:-NFpLsLY5jv8BSO-SpCP:b-gh90ha) ([raw file](https://github.com/work180/auth0-management/blob/c4a31b0718dfa93e741fddffd49694310c3e1b73/src/client.rs#L47)):*
<details><summary><i>Previously, mterring (Michael Terrington) wrote…</i></summary><blockquote>

Sigh... I thought headers returned a string. At least I think remove the `to_owned()` (afiak it's not required) or try this:
</blockquote></details>

Nope, can't compare headers and strings
___
*[`src/client.rs` line 56 at r3](https://reviewable.io/reviews/work180/auth0-management/3#-NFmWr69-WvXd525EcXJ:-NFpM-T5ELQ4JHUOdnau:b-1hroqe) ([raw file](https://github.com/work180/auth0-management/blob/1399cd767c360c0f58e9e0ed0437e3ee0d2e88a4/src/client.rs#L56)):*
<details><summary><i>Previously, mterring (Michael Terrington) wrote…</i></summary><blockquote>

I don't think you want the cost of json encoding the body unless you really need it (which is the not-normal path for this code)
</blockquote></details>

It doesn't like declaring it in the `else` - `borrowed value does not live long enough`

**Review by i-mann — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/work180/auth0-management/3)* status: all files reviewed, 1 unresolved discussion (waiting on @mterring and @robmarto)

___
*[`src/client.rs` line 56 at r3](https://reviewable.io/reviews/work180/auth0-management/3#-NFmWr69-WvXd525EcXJ:-NFqJFDN3oSKbFcgLHDR:b-k054qq) ([raw file](https://github.com/work180/auth0-management/blob/1399cd767c360c0f58e9e0ed0437e3ee0d2e88a4/src/client.rs#L56)):*
<details><summary><i>Previously, i-mann (Isaac Mann) wrote…</i></summary><blockquote>

It doesn't like declaring it in the `else` - `borrowed value does not live long enough`
</blockquote></details>

How's the update?

**Review by mterring — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/work180/auth0-management/3)* status: all files reviewed, 3 unresolved discussions (waiting on @i-mann, @robmarto, and @tonisefton)

___
*[`src/client.rs` line 47 at r2](https://reviewable.io/reviews/work180/auth0-management/3#-NFfVjJP8DtwKcKTMr0E:-NFqYCWj4wgyYXdspriE:b-97lhjj) ([raw file](https://github.com/work180/auth0-management/blob/c4a31b0718dfa93e741fddffd49694310c3e1b73/src/client.rs#L47)):*
<details><summary><i>Previously, i-mann (Isaac Mann) wrote…</i></summary><blockquote>

Nope, can't compare headers and strings
</blockquote></details>

:disappointed:  Can you remove `.to_owned()` from retrieving the headers? I don't think it's needed (it causes the headers to be copied)
___
*[`src/client.rs` line 21 at r4](https://reviewable.io/reviews/work180/auth0-management/3#-NFqZUKDCAZZEjwl5YNM:-NFqZUKDCAZZEjwl5YNN:b-unjprw) ([raw file](https://github.com/work180/auth0-management/blob/c22d87dfade50efb7bfe3eb94c41f074557ed583/src/client.rs#L21)):*
> ```XML
> /// Object for auth0 response for when only a string is returned
> #[derive(Serialize, Deserialize)]
> pub struct Auth0MessageResponse {
> ```

Just noticed this was still here... It's not used anymore
___
*[`src/client.rs` line 56 at r4](https://reviewable.io/reviews/work180/auth0-management/3#-NFqXvbT8LgaynyufAIG:-NFqXvbT8LgaynyufAIH:b-nry2tw) ([raw file](https://github.com/work180/auth0-management/blob/c22d87dfade50efb7bfe3eb94c41f074557ed583/src/client.rs#L56)):*
> ```XML
>       let body = std::str::from_utf8(&body).unwrap();
> 
>       // Note the multiple return points. Different ways were attempted, but
> ```

I'd remove the comment - says more about us than Rust :stuck_out_tongue_winking_eye: - plus given there's already multiple return points, I don't think it's a problem

**Review by i-mann — COMMENTED** [review]
> Reviewed 1 of 4 files at r1, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/work180/auth0-management/3)* status: all files reviewed, 3 unresolved discussions (waiting on @mterring and @robmarto)

___
*[`src/client.rs` line 47 at r2](https://reviewable.io/reviews/work180/auth0-management/3#-NFfVjJP8DtwKcKTMr0E:-NFqqvxDDMg91EtjrHpp:bwgc18h) ([raw file](https://github.com/work180/auth0-management/blob/c4a31b0718dfa93e741fddffd49694310c3e1b73/src/client.rs#L47)):*
<details><summary><i>Previously, mterring (Michael Terrington) wrote…</i></summary><blockquote>

:disappointed:  Can you remove `.to_owned()` from retrieving the headers? I don't think it's needed (it causes the headers to be copied)
</blockquote></details>

Yep. I think that was a layover from a previous version, it's now just programmatic readability for the next line... I've removed the extra call
___
*[`src/client.rs` line 21 at r4](https://reviewable.io/reviews/work180/auth0-management/3#-NFqZUKDCAZZEjwl5YNM:-NFqrKDDEtlKuPerNqxL:bqnyaz3) ([raw file](https://github.com/work180/auth0-management/blob/c22d87dfade50efb7bfe3eb94c41f074557ed583/src/client.rs#L21)):*
<details><summary><i>Previously, mterring (Michael Terrington) wrote…</i></summary><blockquote>

Just noticed this was still here... It's not used anymore
</blockquote></details>

It's used in `mod.rs`
___
*[`src/client.rs` line 56 at r4](https://reviewable.io/reviews/work180/auth0-management/3#-NFqXvbT8LgaynyufAIG:-NFqr5Ye7OJV1r-IzrVR:b-o7uff6) ([raw file](https://github.com/work180/auth0-management/blob/c22d87dfade50efb7bfe3eb94c41f074557ed583/src/client.rs#L56)):*
<details><summary><i>Previously, mterring (Michael Terrington) wrote…</i></summary><blockquote>

I'd remove the comment - says more about us than Rust :stuck_out_tongue_winking_eye: - plus given there's already multiple return points, I don't think it's a problem
</blockquote></details>

Too much trust in Rust!

Done

**Review by mterring — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/work180/auth0-management/3)* status: all files reviewed, 1 unresolved discussion (waiting on @i-mann and @robmarto)

___
*[`src/client.rs` line 21 at r4](https://reviewable.io/reviews/work180/auth0-management/3#-NFqZUKDCAZZEjwl5YNM:-NFz0Mw5DdzKEMWVPLZX:b-ihmosb) ([raw file](https://github.com/work180/auth0-management/blob/c22d87dfade50efb7bfe3eb94c41f074557ed583/src/client.rs#L21)):*
<details><summary><i>Previously, i-mann (Isaac Mann) wrote…</i></summary><blockquote>

It's used in `mod.rs`
</blockquote></details>

That's just an export

**Review by i-mann — COMMENTED** [review]
> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/work180/auth0-management/3)* status: 2 of 4 files reviewed, 1 unresolved discussion (waiting on @mterring and @robmarto)

**Review by i-mann — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/work180/auth0-management/3)* status: 2 of 4 files reviewed, 1 unresolved discussion (waiting on @mterring and @robmarto)

___
*[`src/client.rs` line 21 at r4](https://reviewable.io/reviews/work180/auth0-management/3#-NFqZUKDCAZZEjwl5YNM:-NFzpAp8Cj8obQK9hUBs:bb6un1j) ([raw file](https://github.com/work180/auth0-management/blob/c22d87dfade50efb7bfe3eb94c41f074557ed583/src/client.rs#L21)):*
<details><summary><i>Previously, mterring (Michael Terrington) wrote…</i></summary><blockquote>

That's just an export 
</blockquote></details>

Done

**Review by mterring — APPROVED** [review]
> Reviewed 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/work180/auth0-management/3)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @robmarto)

**Pull request merged**